src/gui/src/renderThread.cpp

```diff
@@ -1306,15 +1306,76 @@ void RenderThread::drawPinMarkers(Painter& painter,
   QPainter* qpainter = static_cast<GuiPainter&>(painter).getPainter();
   const QFont initial_font = qpainter->font();
   QFont marker_font = viewer_->options_->pinMarkersFont();
-  qpainter->setFont(marker_font);
 
-  const QFontMetrics font_metrics(marker_font);
-  // draw names of pins when 100 pins would fit on an edge
-  const bool draw_names
-      = std::max(die_width, die_height) * viewer_->pixels_per_dbu_
-        > 100 * font_metrics.height();
   const int text_margin = 2.0 / viewer_->pixels_per_dbu_;
 
+  const int viewer_width = viewer_->geometry().width();
+  const int viewer_height = viewer_->geometry().height();
+
+  const int gap = (std::min(viewer_width, viewer_height)
+                   - std::max(die_width, die_height) * viewer_->pixels_per_dbu_)
+                  / 2;
+
+  const int available_space
+      = gap - std::ceil(max_dim * viewer_->pixels_per_dbu_);
+  std::vector<int> text_lengths;
+
+  for (auto pin : block->getBTerms()) {
+    text_lengths.push_back(pin->getName().size());
+  }
+
+  auto longest_text_length
+      = std::max_element(text_lengths.begin(), text_lengths.end());
+  int font_size = gap / (*longest_text_length);  // in pixels
+
+  if (font_size <= 0) {
+    font_size = 1;  // to avoid run-time warning when font_size is  <= 0
+  }
+
+  marker_font.setPixelSize(font_size);
+  const QFontMetrics font_metrics(marker_font);
+  std::vector<int> text_widths;
+
+  for (auto pin : block->getBTerms()) {
+    QString pin_text = QString::fromStdString(pin->getName());
+    text_widths.push_back(font_metrics.width(pin_text));
+  }
+
+  auto longest_text_size
+      = std::max_element(text_widths.begin(), text_widths.end());  // pixels
```

Comment:
Building a vector to just take the max is quite wasteful.  Just compute the max as you go and skip all the memory allocations.  Same below.